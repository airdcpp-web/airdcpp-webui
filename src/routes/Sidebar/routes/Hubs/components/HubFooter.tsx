import { useEffect, useState } from 'react';
import { ErrorResponse } from 'airdcpp-apisocket';

import HubConstants from 'constants/HubConstants';

import { useFormatter } from 'context/FormatterContext';
import { SessionFooter, FooterItem } from 'routes/Sidebar/components/SessionFooter';
import EncryptionState from 'components/EncryptionState';

import * as API from 'types/api';
import * as UI from 'types/ui';

import {
  SocketSubscriptionDecorator,
  SocketSubscriptionDecoratorChildProps,
} from 'decorators/SocketSubscriptionDecorator';
import { usingMobileLayout } from 'utils/BrowserUtils';
import { useLayoutWidth } from 'context/LayoutWidthContext';
import { ActionMenu } from 'components/action-menu';
import { HubSettingActionMenu } from 'actions/ui/hub/HubSettingActions';
import { useSocket } from 'context/SocketContext';

interface HubFooterProps {
  session: API.Hub;
  userlistToggle: React.ReactNode;
  sessionT: UI.ModuleTranslator;
}

type DataProps = SocketSubscriptionDecoratorChildProps<HubFooterProps>;
const HubFooter: React.FC<HubFooterProps & DataProps> = (props) => {
  const { formatSize } = useFormatter();
  const socket = useSocket();
  const [counts, setCounts] = useState<API.HubCounts>({
    share_size: 0,
    user_count: 0,
  });

  useEffect(() => {
    // Fetch initial
    socket
      .get<API.HubCounts>(`${HubConstants.SESSIONS_URL}/${props.session.id}/counts`)
      .then(setCounts)
      .catch((error: ErrorResponse) =>
        console.error('Failed to fetch hub counts', error.message),
      );

    // Subscription
    const url = HubConstants.SESSIONS_URL;
    props.addSocketListener(
      url,
      HubConstants.SESSION_COUNTS_UPDATED,
      (data: API.HubCounts) => setCounts(data),
      props.session.id,
    );

    return () => props.removeSocketListeners(props);
  }, [props.session]);

  const layoutWidth = useLayoutWidth();

  const { userlistToggle, session, sessionT } = props;
  const { user_count: users, share_size: shared } = counts;

  const averageShare = formatSize(users > 0 ? shared / users : 0);
  return (
    <SessionFooter>
      <FooterItem
        text={
          <>
            <EncryptionState encryption={session.encryption} />
            {sessionT.t('xUsers', {
              defaultValue: '{{count}} user',
              defaultValue_plural: '{{count}} users',
              count: users,
              replace: {
                count: users,
              },
            })}
          </>
        }
      />
      {!usingMobileLayout(layoutWidth) && (
        <FooterItem
          text={sessionT.t('sharePerUser', {
            defaultValue: '{{total}} ({{average}}/user)',
            replace: {
              total: formatSize(shared),
              average: averageShare,
            },
          })}
        />
      )}
      <div className="actions-wrapper">
        <div className="actions">
          {userlistToggle}
          <ActionMenu
            className="top left pointing"
            actions={HubSettingActionMenu}
            header={sessionT.translate('Notification settings')}
            itemData={session}
            triggerIcon={
              session.settings.use_main_chat_notify || session.settings.show_joins
                ? 'yellow bell'
                : 'bell'
            }
          />
        </div>
      </div>
    </SessionFooter>
  );
};

export default SocketSubscriptionDecorator(
  HubFooter,
  // ({ session }) => !!HubSessionStore.getSession(session.id),
);
