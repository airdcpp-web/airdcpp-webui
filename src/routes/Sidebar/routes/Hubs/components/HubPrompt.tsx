import * as React from 'react';

import ActionInput from 'components/semantic/ActionInput';
import Button from 'components/semantic/Button';

import Icon, { IconType } from 'components/semantic/Icon';

import * as API from 'types/api';
import * as UI from 'types/ui';
import { Trans } from 'react-i18next';
import { sendHubPassword, acceptHubRedirect } from 'services/api/HubApi';
import { runBackgroundSocketAction } from 'utils/ActionUtils';
import IconConstants from 'constants/IconConstants';
import { useSession } from 'context/SessionContext';

interface HubActionPromptProps {
  icon: IconType;
  title: React.ReactNode;
  content: React.ReactNode;
}

// Main prompt (HUBS_EDIT permission is required for the content to be rendered)
const HubActionPrompt: React.FC<HubActionPromptProps> = ({ icon, title, content }) => {
  const { hasAccess } = useSession();
  if (!hasAccess(API.AccessEnum.HUBS_EDIT)) {
    return null;
  }

  return (
    <div className="ui icon message hub-action-prompt">
      <h3 className="ui header">
        <Icon icon={icon} />
        <div className="content">{title}</div>
      </h3>
      {content}
    </div>
  );
};

interface PasswordPromptProps {
  hub: API.Hub;
  sessionT: UI.ModuleTranslator;
}

// Sub prompts
const PasswordPrompt: React.FC<PasswordPromptProps> = ({ hub, sessionT }) => (
  <div className="password prompt">
    <ActionInput
      placeholder={sessionT.translate('Password')}
      caption={sessionT.translate('Submit')}
      icon={IconConstants.CONNECT}
      handleAction={(text) =>
        runBackgroundSocketAction(() => sendHubPassword(hub, text), sessionT.plainT)
      }
    />
    <div className="help">
      <Trans i18nKey={sessionT.toI18nKey('passwordPromptHelp')}>
        This usually means that there's a registered account associated with your nick. If
        you don't remember having a registered account in this hub, there may be someone
        else using the same nick.
      </Trans>
    </div>
  </div>
);

interface RedirectPromptProps {
  hub: API.Hub;
  sessionT: UI.ModuleTranslator;
}

const RedirectPrompt: React.FC<RedirectPromptProps> = ({ hub, sessionT }) => (
  <Button
    className="redirect prompt"
    icon={IconConstants.CONNECT}
    onClick={(_) =>
      runBackgroundSocketAction(() => acceptHubRedirect(hub), sessionT.plainT)
    }
    caption={sessionT.t('acceptRedirect', {
      defaultValue: 'Accept redirect to {{url}}',
      replace: {
        url: hub.connect_state.data!.hub_url,
      },
    })}
  />
);

export { RedirectPrompt, PasswordPrompt, HubActionPrompt };
