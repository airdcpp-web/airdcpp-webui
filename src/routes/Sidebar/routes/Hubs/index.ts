export { default } from './components/Hubs';
