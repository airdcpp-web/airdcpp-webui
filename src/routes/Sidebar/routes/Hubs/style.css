.hub-action-prompt {
  flex-direction: column;
  align-items: flex-start !important;
}

.hub-action-prompt .header {
  margin: 0 0 1rem !important;
}

.hub-action-prompt .password.prompt .help {
  margin-top: 10px;
}

.hub-action-prompt .redirect.prompt {
  max-width: 100%;
  overflow: hidden;
}

.hub.session .virtual-table,
.hub.session .offline-message,
.hub.session .loader-wrapper {
  flex: 1;
  margin-top: 5px;
}

.hub.session .cell-wrapper.tag,
.hub.session .cell-wrapper.nick,
.hub.session .cell-wrapper.description,
.hub.session .cell-wrapper.share_size {
  overflow: hidden;
  white-space: nowrap;
}

.hub.session .cell-wrapper.nick.op {
  color: steelblue;
}
.hub.session .session-footer .actions-wrapper > a,
.hub.session .session-footer .actions-wrapper > div {
  /* Don't cut the shadow */
  padding-left: 1px;
  padding-bottom: 1px;
}

.hub.session .session-footer .userlist-button label {
  color: grey;
  font-size: small;
}

.hub.session .session-footer .actions-wrapper .dropdown {
  margin-left: 10px;
}
