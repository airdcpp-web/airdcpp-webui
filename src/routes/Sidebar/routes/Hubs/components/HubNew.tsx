import HubSearchInput from '@/components/autosuggest/HubSearchInput';
import RecentLayout from '@/routes/Sidebar/components/RecentLayout';

import Message from '@/components/semantic/Message';
import { Link } from 'react-router';

import { HistoryEntryEnum } from '@/constants/HistoryConstants';

import * as API from '@/types/api';

import { Trans, useTranslation } from 'react-i18next';
import IconConstants from '@/constants/IconConstants';
import Button from '@/components/semantic/Button';
import { NewSessionLayoutProps } from '@/routes/Sidebar/components/types';
import LinkButton from '@/components/semantic/LinkButton';
import { HubAPIActions } from '@/actions/store/HubActions';
import { useAppStore } from '@/context/StoreContext';
import { useSocket } from '@/context/SocketContext';

const HubNew: React.FC<NewSessionLayoutProps> = ({ sessionT, navigate }) => {
  const store = useAppStore();
  const { t } = useTranslation();
  const socket = useSocket();

  const handleConnect = (hubUrl: string) => {
    HubAPIActions.createSession(
      { hubUrl },
      {
        store,
        navigate,
        t,
        socket,
      },
    );
  };

  const hasSession = (entry: API.HistoryItem) => {
    return !!store.hubs.getSessionByUrl(entry.hub_url);
  };

  const recentHubRender = (entry: API.HistoryItem) => {
    return (
      <LinkButton onClick={(_) => handleConnect(entry.hub_url)} caption={entry.name} />
    );
  };

  return (
    <div className="hub session new">
      <HubSearchInput
        submitHandler={handleConnect}
        button={
          <Button icon={IconConstants.CONNECT} caption={sessionT.translate('Connect')} />
        }
      />
      <Message
        description={
          <Trans
            i18nKey={sessionT.toI18nKey('favoriteHubsHint')}
            defaults="Tip: visit the <url>Favorite hubs</url> page to connect using custom settings"
            components={{
              url: <Link to="/favorite-hubs" />,
            }}
          />
        }
      />
      <RecentLayout
        entryType={HistoryEntryEnum.HUB}
        hasSession={hasSession}
        entryTitleRenderer={recentHubRender}
        entryIcon="sitemap"
      />
    </div>
  );
};

export default HubNew;
