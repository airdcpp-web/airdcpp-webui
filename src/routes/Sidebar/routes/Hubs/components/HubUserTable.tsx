import PropTypes from 'prop-types';
import * as React from 'react';

import HubSessionStore from 'stores/HubSessionStore';
import HubUserViewStore from 'stores/HubUserViewStore';

import { Column } from 'fixed-data-table-2';
import VirtualTable from 'components/table/VirtualTable';
import { SizeCell, ConnectionCell, IpCell } from 'components/table/Cell';

import { TableUserMenu } from 'components/action-menu';

import Message from 'components/semantic/Message';
import Loader from 'components/semantic/Loader';
import { RowWrapperCellChildProps } from 'components/table/RowWrapperCell';

import * as API from 'types/api';
import * as UI from 'types/ui';
import IconConstants from 'constants/IconConstants';
import MenuConstants from 'constants/MenuConstants';

interface NickCellProps extends RowWrapperCellChildProps<string, API.HubUser> {
  session: API.Hub;
}

const NickCell: React.FC<NickCellProps> = ({ cellData, rowDataGetter, session }) => (
  <TableUserMenu
    text={cellData}
    user={rowDataGetter!()}
    userIcon={true}
    remoteMenuId={MenuConstants.HUB_USER}
    entity={session}
  />
);

interface HubUserTableProps {
  session: API.Hub;
  sessionT: UI.ModuleTranslator;
}

class HubUserTable extends React.Component<HubUserTableProps> {
  static propTypes = {
    session: PropTypes.object, // required
  };

  rowClassNameGetter = (user: API.HubUser) => {
    return user.flags.join(' ');
  };

  emptyRowsNodeGetter = () => {
    const { session, sessionT } = this.props;
    const connectState = session.connect_state.id;

    if (
      connectState === API.HubConnectStateEnum.DISCONNECTED ||
      connectState === API.HubConnectStateEnum.REDIRECT
    ) {
      return (
        <div className="offline-message">
          <Message
            className="offline"
            title={sessionT.translate('Not connected to the hub')}
            icon={IconConstants.OFFLINE}
          />
        </div>
      );
    }

    const text = sessionT.translate(
      connectState !== API.HubConnectStateEnum.CONNECTED
        ? 'Connecting'
        : 'Loading userlist',
    );

    return (
      <div className="loader-wrapper">
        <Loader text={text} />
      </div>
    );
  };

  render() {
    const { session } = this.props;
    return (
      <VirtualTable
        store={HubUserViewStore}
        entityId={session.id}
        sessionStore={HubSessionStore}
        rowClassNameGetter={this.rowClassNameGetter}
        emptyRowsNodeGetter={this.emptyRowsNodeGetter}
        moduleId={UI.Modules.HUBS}
      >
        <Column
          name="Nick"
          width={170}
          columnKey="nick"
          flexGrow={8}
          cell={<NickCell session={session} />}
        />
        <Column
          name="Share size"
          width={85}
          columnKey="share_size"
          cell={<SizeCell />}
          flexGrow={1}
        />
        <Column name="Description" width={100} columnKey="description" flexGrow={1} />
        <Column name="Tag" width={150} columnKey="tag" flexGrow={2} />
        <Column
          name="Upload speed"
          width={80}
          columnKey="upload_speed"
          cell={<ConnectionCell />}
          flexGrow={2}
        />
        <Column
          name="Download speed"
          width={80}
          columnKey="download_speed"
          cell={<ConnectionCell />}
          flexGrow={2}
        />
        <Column
          name="IP (v4)"
          width={120}
          columnKey="ip4"
          flexGrow={3}
          cell={<IpCell />}
        />
        <Column
          name="IP (v6)"
          width={120}
          columnKey="ip6"
          flexGrow={3}
          cell={<IpCell />}
        />
        <Column name="Files" width={70} columnKey="file_count" flexGrow={3} />
      </VirtualTable>
    );
  }
}

export default HubUserTable;
