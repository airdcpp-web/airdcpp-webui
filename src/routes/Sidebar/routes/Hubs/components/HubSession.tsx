import { useEffect, useState } from 'react';

import HubMessageStore from 'stores/HubMessageStore';

import { loadSessionProperty, saveSessionProperty } from 'utils/BrowserUtils';
import Checkbox from 'components/semantic/Checkbox';

import ChatLayout from 'routes/Sidebar/components/chat/ChatLayout';
import HubUserTable from 'routes/Sidebar/routes/Hubs/components/HubUserTable';

import HubFooter from 'routes/Sidebar/routes/Hubs/components/HubFooter';
import {
  RedirectPrompt,
  PasswordPrompt,
  HubActionPrompt,
} from 'routes/Sidebar/routes/Hubs/components/HubPrompt';

import '../style.css';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { shareTempFile } from 'services/api/ShareApi';
import HubActions from 'actions/reflux/HubActions';
import IconConstants from 'constants/IconConstants';
import MenuConstants from 'constants/MenuConstants';
import { SessionChildProps } from 'routes/Sidebar/components/types';
import { useSession } from 'context/SessionContext';

const getStorageKey = (sessionId: number) => {
  return `view_userlist_${sessionId}`;
};

const checkList = (sessionId: number) => {
  const value = loadSessionProperty(getStorageKey(sessionId), false);
  return value;
};

type HubSessionProps = SessionChildProps<API.Hub, UI.EmptyObject, UI.ChatActionList>;

const HubSession: React.FC<HubSessionProps> = ({
  session,
  sessionApi,
  sessionT,
  uiActions,
}) => {
  const [showList, setShowList] = useState(checkList(session.id));
  const login = useSession();

  const toggleListState = () => {
    setShowList(!showList);
    return !showList;
  };

  useEffect(() => {
    setShowList(checkList(session.id));
  }, [session.id]);

  const getMessage = () => {
    const connectState = session.connect_state.id;

    if (connectState === API.HubConnectStateEnum.PASSWORD) {
      return (
        <HubActionPrompt
          title={sessionT.translate('Password required')}
          icon={IconConstants.LOCK}
          content={<PasswordPrompt hub={session} sessionT={sessionT} />}
        />
      );
    }

    if (connectState === API.HubConnectStateEnum.REDIRECT) {
      return (
        <HubActionPrompt
          title={sessionT.translate('Redirect requested')}
          icon={IconConstants.REDIRECT}
          content={<RedirectPrompt hub={session} sessionT={sessionT} />}
        />
      );
    }

    return null;
  };

  const onClickUsers = () => {
    const newValue = toggleListState();
    saveSessionProperty(getStorageKey(session.id), newValue);
  };

  const handleFileUpload = (file: File) => {
    const { hub_url } = session;
    return shareTempFile(file, hub_url, undefined, login);
  };

  const checkbox = (
    <Checkbox
      id="userlist-toggle"
      type="toggle"
      caption={sessionT.translate('User list')}
      onChange={onClickUsers}
      checked={showList}
    />
  );

  return (
    <div className="hub chat session">
      {getMessage()}
      {showList ? (
        <HubUserTable session={session} sessionT={sessionT} />
      ) : (
        <ChatLayout
          messageStore={HubMessageStore}
          chatApi={HubActions as UI.ChatAPI}
          sessionApi={sessionApi}
          chatActions={uiActions}
          chatAccess={API.AccessEnum.HUBS_SEND}
          session={session}
          handleFileUpload={handleFileUpload}
          highlightRemoteMenuId={MenuConstants.HUB_MESSAGE_HIGHLIGHT}
          hubUrl={session.hub_url}
        />
      )}
      <HubFooter userlistToggle={checkbox} session={session} sessionT={sessionT} />
    </div>
  );
};

export default HubSession;
