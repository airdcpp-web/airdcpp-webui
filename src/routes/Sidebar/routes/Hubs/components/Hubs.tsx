import * as React from 'react';

import { TextDecorator } from 'components/text';

import SessionLayout from 'routes/Sidebar/components/SessionLayout';

import HubSessionStore from 'stores/HubSessionStore';
import HubAPIActions from 'actions/reflux/HubActions';

import { hubOnlineStatusToColor } from 'utils/TypeConvert';

import HubSession from 'routes/Sidebar/routes/Hubs/components/HubSession';
import HubNew from 'routes/Sidebar/routes/Hubs/components/HubNew';
import HubIcon from 'components/icon/HubIcon';

import * as API from 'types/api';
import * as UI from 'types/ui';
import {
  SessionProviderDecorator,
  SessionProviderDecoratorChildProps,
} from 'routes/Sidebar/decorators/SessionProviderDecorator';
import IconConstants from 'constants/IconConstants';
import { Params } from 'react-router-dom';
import { HubActionMenu } from 'actions/ui/hub';

const ItemHandler: UI.SessionInfoGetter<API.Hub> = {
  itemNameGetter(session) {
    return session.identity.name;
  },

  itemStatusGetter(session) {
    return hubOnlineStatusToColor(session.connect_state.id);
  },

  itemHeaderDescriptionGetter(session) {
    return <TextDecorator text={session.identity.description} emojify={true} />;
  },

  itemHeaderIconGetter(session) {
    return <HubIcon hub={session} />;
  },
};

const parseNumericId = (params: Readonly<Params<string>>) => {
  if (!params.id) {
    return undefined;
  }

  return parseInt(params['id']);
};
const sessionActions = ['reconnect', 'clearChat', 'favorite'];

const Hubs: React.FC<SessionProviderDecoratorChildProps<API.Hub>> = (props) => {
  const { params, sessionT, ...other } = props;
  return (
    <SessionLayout
      activeId={parseNumericId(params)}
      baseUrl="hubs"
      newCaption={sessionT.t('new', 'Connect')}
      newDescription={sessionT.t('newDesc', 'Connect to a new hub')}
      newIcon={IconConstants.HUBS_PLAIN}
      editAccess={API.AccessEnum.HUBS_EDIT}
      uiActions={HubActionMenu}
      actionIds={sessionActions}
      sessionApi={HubAPIActions as UI.SessionActions<API.Hub>}
      sessionItemLayout={HubSession}
      newLayout={HubNew}
      sessionT={sessionT}
      unreadInfoStore={HubSessionStore}
      {...ItemHandler}
      {...other}
    />
  );
};

export default SessionProviderDecorator(Hubs, HubSessionStore, UI.Modules.HUBS);
