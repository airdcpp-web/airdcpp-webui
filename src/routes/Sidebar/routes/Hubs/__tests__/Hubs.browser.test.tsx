import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { waitFor, within } from '@testing-library/react';

import { renderDataRoutes } from '@/tests/render/test-renderers';

import { CommonDataMocks, initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import { setupUserEvent, toggleCheckboxValue } from '@/tests/helpers/test-form-helpers';

import * as API from '@/types/api';

import SocketNotificationListener from '@/components/main/notifications/SocketNotificationListener';

import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';
/*import {
  waitExpectRequestToMatchSnapshot,
  navigateToUrl,
  waitForData,
  waitForUrl,
} from '@/tests/helpers/test-helpers';*/
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import { getHistoryUrl } from '@/routes/Sidebar/components/RecentLayout';
import HistoryConstants, { HistoryEntryEnum } from '@/constants/HistoryConstants';

import { installActionMenuMocks } from '@/components/action-menu/__tests__/test-action-menu-helpers';
import MenuConstants from '@/constants/MenuConstants';
import { RemoteMenuGrouped1 } from '@/tests/mocks/api/menu';

import { createDataFetchRoutes } from '@/tests/helpers/test-route-helpers';
import {
  HubADC1,
  HubADC2,
  HubADC3,
  HubCountsResponse,
  HubNMDC1,
} from '@/tests/mocks/api/hubs';
import HubConstants from '@/constants/HubConstants';
import {
  installSessionMessageMocks,
  waitSessionsLoaded,
} from '@/tests/mocks/mock-session';
import { HistoryHubResponse, HistoryHubSearchResponse } from '@/tests/mocks/api/history';
import Hubs from '../components/Hubs';
import { installTableMocks } from '@/tests/mocks/mock-table';
import {
  Hub1MessageMe,
  HubADC1MessageListResponse,
  HubDisconnectedMessageListResponse,
} from '@/tests/mocks/api/hub-messages';
import { Hub1Bot, Hub1UserListResponse } from '@/tests/mocks/api/user';
import {
  navigateToUrl,
  waitExpectRequestToMatchSnapshot,
  waitForData,
  waitForUrl,
} from '@/tests/helpers/test-helpers';

// tslint:disable:no-empty
describe('Hubs', () => {
  let server: MockServer;

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  const installHubSessionMocks = (sessionId: number) => {
    server.addRequestHandler(
      'GET',
      `${HubConstants.SESSIONS_URL}/${sessionId}/counts`,
      HubCountsResponse,
    );

    const countsListener = server.addSubscriptionHandler(
      HubConstants.MODULE_URL,
      HubConstants.SESSION_COUNTS_UPDATED,
      sessionId,
    );

    return { countsListener };
  };

  const installNewLayoutMocks = (commonData: CommonDataMocks) => {
    server.addRequestHandler(
      'GET',
      getHistoryUrl(HistoryEntryEnum.HUB),
      HistoryHubResponse,
    );

    server.addRequestHandler(
      'POST',
      `${HistoryConstants.SESSIONS_URL}/${HistoryEntryEnum.HUB}/search`,
      HistoryHubSearchResponse,
    );

    const onSessionCreated = vi.fn(() => {
      commonData.mockStoreListeners.hub.created.fire(HubADC1);
    });

    server.addRequestHandler(
      'POST',
      HubConstants.SESSIONS_URL,
      HubADC1,
      onSessionCreated,
    );

    return { onSessionCreated };
  };

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server, {
      permissions: [
        API.AccessEnum.HUBS_VIEW,
        API.AccessEnum.HUBS_EDIT,
        API.AccessEnum.HUBS_SEND,
      ],
    });

    // Messages
    const hub1MessageMocks = installSessionMessageMocks(
      HubConstants.SESSIONS_URL,
      HubADC1.id,
      HubADC1MessageListResponse,
      server,
    );

    const hub2MessageMocks = installSessionMessageMocks(
      HubConstants.SESSIONS_URL,
      HubADC2.id,
      HubDisconnectedMessageListResponse,
      server,
    );

    // New layout
    const { onSessionCreated } = installNewLayoutMocks(commonData);

    // Menus
    const menuMocks = installActionMenuMocks(
      MenuConstants.HUB_MESSAGE_HIGHLIGHT,
      [RemoteMenuGrouped1],
      server,
    );

    // Tables
    const hub1UserTableMocks = installTableMocks(Hub1UserListResponse, {
      server,
      moduleUrl: HubConstants.MODULE_URL,
      viewName: HubConstants.USER_VIEW_ID,
      entityId: HubADC1.id,
    });

    const hub2UserTableMocks = installTableMocks([], {
      server,
      moduleUrl: HubConstants.MODULE_URL,
      viewName: HubConstants.USER_VIEW_ID,
      entityId: HubADC2.id,
    });

    // Generic
    const sessionMocks1 = installHubSessionMocks(HubADC1.id);
    const sessionMocks2 = installHubSessionMocks(HubADC2.id);

    return {
      commonData,

      hub1Mocks: {
        ...hub1MessageMocks,
        ...sessionMocks1,
        userTable: hub1UserTableMocks,
      },

      hub2Mocks: {
        ...hub2MessageMocks,
        ...sessionMocks2,
        userTable: hub2UserTableMocks,
      },

      onSessionCreated,

      menuMocks,
    };
  };

  const renderLayout = async (userActive = true) => {
    const { commonData, ...other } = await getSocket();

    if (userActive) {
      commonData.sessionStore.getState().activity.setUserActive(true);
    }

    const HubLayoutTest = () => {
      return (
        <>
          <Hubs />
          <SocketNotificationListener />
        </>
      );
    };

    const routes = createDataFetchRoutes([
      {
        path: '/hubs/:session?/:id?/*',
        Component: HubLayoutTest,
      },
    ]);

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/hubs'] },
      viewType: VIEW_FIXED_HEIGHT,
    });

    const userEvent = setupUserEvent();
    return { ...commonData, ...renderData, ...other, userEvent };
  };

  test('should render', async () => {
    const renderResult = await renderLayout();
    const { getByText, queryByText, router } = renderResult;

    await waitSessionsLoaded(queryByText);

    // Check messages
    await waitFor(() => expect(getByText(Hub1MessageMe.text)).toBeTruthy());

    // Open user list
    await toggleCheckboxValue('User list', renderResult);

    await waitFor(() => expect(getByText(Hub1Bot.nick)).toBeTruthy());

    await navigateToUrl('/', router);
  });

  test('open new session', async () => {
    const {
      queryByText,
      sessionStore,
      router,
      getByRole,
      findByText,
      getByText,
      userEvent,
    } = await renderLayout();

    await waitSessionsLoaded(queryByText);

    // Remove existing sessions
    sessionStore.getState().hubs.removeSession(HubADC1);
    sessionStore.getState().hubs.removeSession(HubADC2);
    sessionStore.getState().hubs.removeSession(HubADC3);
    sessionStore.getState().hubs.removeSession(HubNMDC1);

    // We should be redirected to the new session layout
    await waitForUrl('/hubs/new', router);
    await waitForData('Loading data...', queryByText);

    // Type the hub name
    const combobox = getByRole('combobox');
    const input = within(combobox).getByRole('textbox');
    await userEvent.type(input, HubADC1.identity.name);

    // Wait until suggestion appears, then select via keyboard
    await findByText(HubADC1.identity.name);
    await userEvent.keyboard('{ArrowDown}{Enter}');

    // Click the hub
    // const hub1Item = getByText(HubADC1.identity.name);
    // await userEvent.click(hub1Item!);

    // We should be redirected to the new session
    await waitFor(() =>
      expect(sessionStore.getState().hubs.activeSessionId).toEqual(HubADC1.id),
    );

    await waitFor(() => expect(getByText(Hub1MessageMe.text)).toBeTruthy());
  });

  describe('prompts', () => {
    test('should handle redirect prompt', async () => {
      const { queryByText, getByText, mockStoreListeners, findByText, userEvent } =
        await renderLayout(false);

      const onRedirect = vi.fn();
      server.addRequestHandler(
        'POST',
        `${HubConstants.SESSIONS_URL}/${HubADC1.id}/redirect`,
        HubADC1,
        onRedirect,
      );

      await waitSessionsLoaded(queryByText);

      // Fire the redirect prompt event
      mockStoreListeners.hub.updated.fire(
        {
          connect_state: {
            data: {
              hub_url: 'adcs://test_redirect.org:12345',
            },
            id: 'redirect',
            str: 'Redirect',
          },
          encryption: null,
        },
        HubADC1.id,
      );

      await findByText('Redirect requested');

      // Accept it
      await userEvent.click(
        getByText('Accept redirect to adcs://test_redirect.org:12345'),
      );

      // Check that the request was sent
      await waitExpectRequestToMatchSnapshot(onRedirect);
    });

    test('should handle password prompt', async () => {
      const {
        findByText,
        getByText,
        queryByText,
        mockStoreListeners,
        getByPlaceholderText,
        userEvent,
      } = await renderLayout();

      const onPassword = vi.fn();
      server.addRequestHandler(
        'POST',
        `${HubConstants.SESSIONS_URL}/${HubADC2.id}/password`,
        HubADC1,
        onPassword,
      );

      await waitSessionsLoaded(queryByText);

      // Switch to the disconnected session
      const session2MenuItem = queryByText(HubADC2.identity.name);
      await userEvent.click(session2MenuItem!);

      await findByText('Connection timeout');

      // Fire the password prompt event
      mockStoreListeners.hub.updated.fire(
        {
          connect_state: {
            id: 'password',
            str: 'Password requested',
          },
        },
        HubADC2.id,
      );

      await findByText('Password required');

      // Fill in the password
      const passwordInput = getByPlaceholderText('Password');
      await userEvent.type(passwordInput, 'testpassword');

      await userEvent.click(getByText('Submit'));

      // Check that the request was sent
      await waitExpectRequestToMatchSnapshot(onPassword);
    });
  });
});
