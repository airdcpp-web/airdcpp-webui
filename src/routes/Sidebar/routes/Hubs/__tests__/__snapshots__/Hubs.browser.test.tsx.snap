// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hubs > prompts > should handle password prompt 1`] = `
{
  "data": {
    "password": "testpassword",
  },
  "method": "POST",
  "path": "hubs/2/password",
}
`;

exports[`Hubs > prompts > should handle redirect prompt 1`] = `
{
  "method": "POST",
  "path": "hubs/1/redirect",
}
`;
